expression/builtin_cast.go

```diff
@@ -467,9 +468,79 @@ func (b *castJSONAsArrayFunctionSig) evalJSON(row chunk.Row) (res types.BinaryJS
 		return types.BinaryJSON{}, false, ErrNotSupportedYet.GenWithStackByArgs("CAST-ing Non-JSON Array type to array")
 	}
 
-	// TODO: impl the cast(... as ... array) function
+	arrayVals := make([]any, 0, len(b.args))
+	f := convertJSON2Tp(b.tp.ArrayType())
+	if f == nil {
+		return types.BinaryJSON{}, false, ErrNotSupportedYet.GenWithStackByArgs("CAS-ing JSON to the target type")
+	}
+	sc := b.ctx.GetSessionVars().StmtCtx
+	originalOverflowAsWarning := sc.OverflowAsWarning
+	originIgnoreTruncate := sc.IgnoreTruncate
+	originTruncateAsWarning := sc.TruncateAsWarning
+	sc.OverflowAsWarning = false
+	sc.IgnoreTruncate = false
+	sc.TruncateAsWarning = false
+	defer func() {
+		sc.OverflowAsWarning = originalOverflowAsWarning
+		sc.IgnoreTruncate = originIgnoreTruncate
+		sc.TruncateAsWarning = originTruncateAsWarning
+	}()
+	for i := 0; i < val.GetElemCount(); i++ {
+		item, err := f(sc, val.ArrayGetElem(i))
+		if err != nil {
+			return types.BinaryJSON{}, false, err
+		}
+		arrayVals = append(arrayVals, item)
+	}
+	return types.CreateBinaryJSON(arrayVals), false, nil
+}
 
-	return types.BinaryJSON{}, false, nil
+func convertJSON2Tp(tp *types.FieldType) func(*stmtctx.StatementContext, types.BinaryJSON) (any, error) {
+	switch tp.EvalType() {
+	case types.ETString:
+		return func(sc *stmtctx.StatementContext, item types.BinaryJSON) (any, error) {
+			if item.TypeCode != types.JSONTypeCodeString {
+				return nil, ErrInvalidJSONForFuncIndex
+			}
+			return types.ProduceStrWithSpecifiedTp(string(item.GetString()), tp, sc, false)
+		}
+	case types.ETInt:
+		return func(sc *stmtctx.StatementContext, item types.BinaryJSON) (any, error) {
+			if item.TypeCode != types.JSONTypeCodeInt64 && item.TypeCode != types.JSONTypeCodeUint64 {
+				return nil, ErrInvalidJSONForFuncIndex
+			}
+			return types.ConvertJSONToInt(sc, item, mysql.HasUnsignedFlag(tp.GetFlag()), tp.GetType())
+		}
+	case types.ETReal, types.ETDecimal:
+		return func(sc *stmtctx.StatementContext, item types.BinaryJSON) (any, error) {
+			if item.TypeCode != types.JSONTypeCodeInt64 && item.TypeCode != types.JSONTypeCodeUint64 && item.TypeCode != types.JSONTypeCodeFloat64 {
+				return nil, ErrInvalidJSONForFuncIndex
+			}
+			return types.ConvertJSONToFloat(sc, item)
+		}
+	case types.ETDatetime:
```

Comment:
Could we also support `case mysql.TypeTimestamp`?

Reply:
`timestamp` is not able to be the target type of the cast function